_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+
+```
+// Delete Users
+// DELETE: /Users/:userId
+// RFC Notes on Deleting Users: https://www.rfc-editor.org/rfc/rfc7644#section-3.6 
+ scimRoute.delete('/Users/:userId', passport.authenticate('bearer'), async (req, res) => {
+    // console.log('DELETE: /Users/:userId'); 
+    // console.log(req.body);
+  
+     const id  = parseInt(req.params.userId);
+     await prisma.user.delete({
+       where: { id }
+     });
+  
+     res.sendStatus(204);
+   });
+```
+##### Test deleting a user 
+Test out the delete function by making a DELETE request with an empty body to `http://localhost:3333/scim/v2/Users/3`.
+
+What response do you expect? You should get an HTTP 204 with no body in response.
+
+
+##### Test soft deleting/"deprovisioning" a user
+
+Okta implements a [soft delete aka deprovision](https://developer.okta.com/docs/reference/scim/scim-20/#delete-users) via PUT or PATCH. This will cover the scenarios such as users reassignment from services and potentially auditing purposes. Add the following code to `scim.ts` to support what the spec calls a partial delete:
+
+```
+// Soft Delete Users
+// PATCH: /Users/:userId
+// RFC Notes on Partial Update: https://www.rfc-editor.org/rfc/rfc7644#section-3.5.2 
+// Note: this does not a true "delete"; this will update the active flag to false (this is an Okta best practice)
+scimRoute.patch('/Users/:userId', passport.authenticate('bearer'), async (req, res) => {
+    // console.log('PATCH: /Users/:userId'); 
+    console.log(req.body);
+  
+     const id  = parseInt(req.params.userId);
+     const active = req.body["Operations"][0]["value"]["active"]
+     await prisma.user.update({
+       data: {
+        active
+       },
+       where: { id }
+     });
+  
+     res.sendStatus(204);
+   });   
+```
+
+Try sending a PATCH request with an empty body to `http://localhost:3333/scim/v2/Users/2`.
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:api:messages:2.0:PatchOp"
+    ],
+    "Operations": [
+        {
+            "op": "replace",
+            "value": {
+                "active": false
+            }
+        }
+    ]
+}
+```
+
+Just like the delete request, you should expect to receive a 204 response; this is in line with [Okta's docs](https://developer.okta.com/docs/reference/scim/scim-20/#update-a-specific-user-patch)
+
+## Connecting with an Identity provider
+
+Now that we have a working SCIM application, let's test it with an Identity Provider! In this workshop, we're using Okta. 
+
+We'll need to set up a few things for Okta to authenticate and interact with our local server. 
+
+### 1) Give your app a public URL
+One way to give your app a public URL or IP would be to host it on a cloud instance with DNS that you control. But if that isn't an option, you can use Localtunnel to temporarily issue a public web address to your own computer. 
```

Comment:
The goal is to have your production app expose these endpoints so that the IDP can make requests to your app's SCIM endpoints. "if that isn't an option" sounds weird, just say that for development purposes yo ucan use localtunnel to provide a public address to the app running on your own computer.